File: tests/cpp/test_chord.cpp
// Gingo — Chord module tests
// SPDX-License-Identifier: MIT

#include <catch2/catch_test_macros.hpp>
#include <gingo/chord.hpp>

#include <algorithm>
#include <string>
#include <vector>

using namespace gingo;

static std::vector<std::string> note_names(const std::vector<Note>& notes) {
    std::vector<std::string> names;
    for (const auto& n : notes) names.push_back(n.natural());
    return names;
}

TEST_CASE("Major triad", "[chord]") {
    Chord c("CM");
    REQUIRE(c.root().natural() == "C");
    REQUIRE(c.type() == "M");
    auto names = note_names(c.notes());
    REQUIRE(names == std::vector<std::string>{"C", "E", "G"});
}

TEST_CASE("Minor seventh chord", "[chord]") {
    Chord c("Am7");
    REQUIRE(c.root().natural() == "A");
    REQUIRE(c.type() == "m7");
    REQUIRE(c.size() == 4);
    auto labels = c.interval_labels();
    REQUIRE(labels == std::vector<std::string>{"P1", "3m", "5J", "7m"});
}

TEST_CASE("Chord with flat root", "[chord]") {
    Chord c("Dbm7");
    REQUIRE(c.root().natural() == "C#");
    auto names = note_names(c.notes());
    REQUIRE(names == std::vector<std::string>{"C#", "E", "G#", "B"});
}

TEST_CASE("Chord with sharp root", "[chord]") {
    Chord c("A#m7");
    auto names = note_names(c.notes());
    REQUIRE(names == std::vector<std::string>{"A#", "C#", "F", "G#"});
}

TEST_CASE("Diminished chord", "[chord]") {
    Chord c("Bdim");
    REQUIRE(c.type() == "dim");
    auto labels = c.interval_labels();
    REQUIRE(labels == std::vector<std::string>{"P1", "3m", "d5"});
}

TEST_CASE("Augmented chord", "[chord]") {
    Chord c("Caug");
    auto labels = c.interval_labels();
    REQUIRE(labels == std::vector<std::string>{"P1", "3M", "#5"});
}

TEST_CASE("Dominant seventh", "[chord]") {
    Chord c("G7");
    REQUIRE(c.type() == "7");
    REQUIRE(c.size() == 4);
}

TEST_CASE("Chord contains note", "[chord]") {
    Chord c("CM");
    REQUIRE(c.contains(Note("E")));
    REQUIRE(c.contains(Note("G")));
    REQUIRE_FALSE(c.contains(Note("F")));
}

TEST_CASE("Identify chord from notes", "[chord]") {
    auto c = Chord::identify({"C", "E", "G"});
    REQUIRE(c.type() == "M");
    REQUIRE(c.root().natural() == "C");
}

TEST_CASE("Chord equality", "[chord]") {
    REQUIRE(Chord("CM") == Chord("CM"));
    REQUIRE(Chord("CM") != Chord("Cm"));
}

File: tests/python/test_chord.py
# Gingo — Chord tests (Python)
# SPDX-License-Identifier: MIT

import pytest
from gingo import Chord, Note


def test_major_triad():
    c = Chord("CM")
    assert c.root().natural() == "C"
    assert c.type() == "M"
    names = [n.natural() for n in c.notes()]
    assert names == ["C", "E", "G"]


def test_minor_seventh():
    c = Chord("Am7")
    assert c.size() == 4
    assert c.interval_labels() == ["P1", "3m", "5J", "7m"]


def test_flat_root():
    c = Chord("Dbm7")
    names = [n.natural() for n in c.notes()]
    assert names == ["C#", "E", "G#", "B"]


def test_dominant_seventh():
    c = Chord("G7")
    assert c.type() == "7"
    assert c.size() == 4


def test_contains():
    c = Chord("CM")
    assert c.contains(Note("E"))
    assert not c.contains(Note("F"))


def test_identify_from_strings():
    c = Chord.identify(["C", "E", "G"])
    assert c.root().natural() == "C"
    assert "M" in c.type() or c.type() == "M"


def test_equality():
    assert Chord("CM") == Chord("CM")
    assert Chord("CM") != Chord("Cm")


def test_string_repr():
    c = Chord("Cm7")
    assert "Cm7" in str(c)
